File: fondamenti-di-programmazione-2/esercizi/primes.cpp
#include <cmath>
#include <iostream>

bool isPrime(int);

int main() {
  int n;

  std::cout << "Scrivi un numero: ";
  if (std::cin >> n)
    if (isPrime(n))
      std::cout << n << " è primo." << std::endl;
    else
      std::cout << n << " non è primo." << std::endl;
  else {
    std::cerr << "Il numero che hai inserito non è valido." << std::endl;
    return -1;
  }

  return 0;
}

bool isPrime(int n) {
  if (n <= 1)
    return false;
  else if (n < 4) // 2 e 3 sono primi
    return true;
  else if (n % 2 == 0) // i numeri pari non sono primi
    return false;
  else if (n < 9)
    return true;
  else if (n % 3 == 0) // i multipli di 3 non sono primi
    return false;
  else { // ciò che resta
    int limite = floor(sqrt(n));
    int divisore = 5;

    while (divisore <= limite) {
      if (n % divisore == 0 || n % (divisore + 2) == 0)
        return false;

      divisore += 6;
    }
  }

  return true;
}

File: Fondamenti di programmazione 1/primes.py
from math import floor, sqrt
n = int(input("Scrivi un numero: "))

def isPrime(n):
    if n == 1 or n == 0:
        return False
    elif n < 4:
        return True # 2 e 3 sono primi
    elif n % 2 == 0:
        return False # i pari non sono primi (eccetto 2)
    elif n < 9:
        return True
    elif n % 3 == 0: # i multipli di 3 non sono primi
        return False
    else: # ciò che resta
        limite = floor(sqrt(n))
        divisore = 5

        while divisore <= limite:
            if n % divisore == 0 or n % (divisore + 2) == 0:
                return False

            divisore += 6

    return True

if isPrime(n):
    print(n, "è primo.")
else:
    print(n, "non è primo.")
